### PR TITLE
[Merged by Bors] - feat(RingTheory/*): prerequisites to the locality of projectivity

### PR DESCRIPTION
**Pull request by erdOne**
Partially written at the AIM algebraic geometry workshop

Co-authored-by: David Swinarski

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Module/Submodule/Localization.lean [hunk 1]
```diff
@@ -57,6 +57,28 @@ lemma Submodule.mem_localized' (x : N) :
 abbrev Submodule.localized : Submodule (Localization p) (LocalizedModule p M) :=
   M'.localized' (Localization p) p (LocalizedModule.mkLinearMap p M)
 
+@[simp]
+lemma Submodule.localized'_bot : (⊥ : Submodule R M).localized' S p f = ⊥ := by
+  rw [← le_bot_iff]
+  rintro _ ⟨_, rfl, s, rfl⟩
+  simp only [IsLocalizedModule.mk'_zero, mem_bot]
```
Comment (chrisflav) [inline Mathlib/Algebra/Module/Submodule/Localization.lean:64]: While you are at it, could you please also add `Submodule.localized'_top`?

File: Mathlib/RingTheory/Localization/Module.lean [hunk 2]
```diff
@@ -213,6 +218,70 @@ def LinearMap.extendScalarsOfIsLocalization (f : M →ₗ[R] N) : M →ₗ[A] N
...
+  ext x
+  induction x using induction_on
+  simp only [LocalizedModule.map_mk, LinearMap.id_coe, id_eq]
+
+lemma LocalizedModule.map_surjective (l : M →ₗ[R] N) (hl : Function.Surjective l) :
+    Function.Surjective (map S l) := by
+  intro x
+  induction' x using LocalizedModule.induction_on with m s
+  obtain ⟨m, rfl⟩ := hl m
+  exact ⟨mk m s, map_mk _ _ _ _⟩
```
[the remark this answers is not in the document]
Comment (chrisflav) [reply Mathlib/RingTheory/Localization/Module.lean:285]: Or even for `IsLocalizedModule.map`? Maybe you can even extract `IsLocalizedModule.map_surjective` (and `IsLocalizedModule.map_injective` from `IsLocalizedModule.map_exact`?

[the remark this answers is not in the document]
Comment (erdOne) [reply Mathlib/RingTheory/Localization/Module.lean]: I proved them for `IsLocalizedModule.map`. It's easier to prove them directly than through `map_exact`.

File: Mathlib/Algebra/Module/LocalizedModule.lean [hunk 3]
```diff
@@ -1146,6 +1171,13 @@ theorem map_exact (g : M₀ →ₗ[R] M₁) (h : M₁ →ₗ[R] M₂) (ex : Func
   Function.Exact.of_ladder_linearEquiv_of_exact
     (map_iso_commute S f₀ f₁ g) (map_iso_commute S f₁ f₂ h) (LocalizedModule.map_exact S g h ex)
 
+/-- Localization of composition is the composition of localization -/
+theorem map_comp' (g : M₀ →ₗ[R] M₁) (h : M₁ →ₗ[R] M₂) :
+    map S f₀ f₂ (h ∘ₗ g) = map S f₁ f₂ h ∘ₗ map S f₀ f₁ g := by
```
Comment (chrisflav) [inline Mathlib/Algebra/Module/LocalizedModule.lean:1176]: Should this be called `map_comp_map` as in `IsLocalization.map_comp_map`? Also `IsLocalization.map_comp_map` is stated the other way round, but your version looks like the correct order. So I guess we keep this order and remember to fix `IsLocalization.map_comp_map` one day?

**Review by chrisflav — APPROVED** [review]
Thanks! Modulo one comment, LGTM!

File: Mathlib/Algebra/Module/LocalizedModule.lean (hunk 3 above)
Comment (chrisflav) [reply Mathlib/Algebra/Module/LocalizedModule.lean:1176]: Oh that makes sense. Then I guess it is `map_comp` that should be named `map_comp_algebraMap` instead, but not in this PR.

**Comment by chrisflav** [comment]
Is there anything we can do about the large import warning? It's only barely above the threshold, but since Kim made it clear on #18134 that we should be more careful with these, let's have a look.

File: Mathlib/RingTheory/Localization/Module.lean [hunk 4]
```diff
@@ -117,6 +117,11 @@ theorem Basis.ofIsLocalizedModule_span :
     _ = map f (span R (Set.range b)) := by rw [Submodule.map_span]
     _ = LinearMap.range f := by rw [b.span_eq, Submodule.map_top]
 
+include S f in
+theorem Module.free_of_isLocalizedModule [Module.Free R M] : Module.Free Rₛ Mₛ :=
+  Module.Free.of_basis
+    (Basis.ofIsLocalizedModule Rₛ S f (Module.Free.chooseBasis R M))
+
```
Comment (chrisflav) [inline Mathlib/RingTheory/Localization/Module.lean:124]: I guess it is only this declaration causing the large-import (that I asked you to move here from `Algebra.Module.Submodule.Localization` to make it more discoverable), so maybe simply revert that move? Sorry for the back and forth.

**Comment by chrisflav** [comment]
Thanks!

maintainer merge

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**